op-acceptance-tests/tests/base/chain/unsafe_sync_test.go

```diff
@@ -0,0 +1,86 @@
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-devstack/devtest"
+	"github.com/ethereum-optimism/optimism/op-devstack/presets"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+	"github.com/stretchr/testify/require"
+)
+
+// TestUnsafeSync tests that the verifier nodes are able to sync the unsafe head of the chain.
+// It does this by waiting for the sequencer to produce 10 unsafe blocks and requiring the verifier nodes to
+// be no more than 10 blocks behind the unsafe head.
+func TestUnsafeSync(gt *testing.T) {
+	t := devtest.SerialT(gt)
+	sys := presets.NewSingleChainMultiNode(t)
+
+	// Stop batcher so there is no sync from deriving from L1
+	// This won't work on external networks.
+	sys.L2Batcher.Stop()
+
+	l2chainID := sys.L2Chain.Escape().ChainID()
+	const NUM_UNSAFE_BLOCKS = 3
+	const TIMEOUT = 5 * time.Second // time to wait for all verifier nodes to catch up to the sequencer
+
+	// In order for this test to be valid, we need at least 2 L2 EL nodes.
+	t.Require().Greater(len(sys.L2Chain.L2ELNodes()), 1, "expected at least 2 L2 EL nodes")
+
+	// Find the index of the sequencer node.
+	indexOfSequencer := 0
+	for i, el := range sys.L2Chain.L2ELNodes() {
+		if strings.Contains(el.ID().String(), "sequencer") {
+			indexOfSequencer = i
+			break
+		}
+	}
+
+	// Wait for the sequencer to produce at least NUM_UNSAFE_BLOCKS unsafe blocks.
+	initialUnsafeHeadNumber := sys.L2Chain.L2ELNodes()[0].ChainSyncStatus(sys.L2Chain.Escape().ChainID(), types.LocalUnsafe).Number
+	finalUnsafeHeadNumber := initialUnsafeHeadNumber + NUM_UNSAFE_BLOCKS
+	ticker := time.NewTicker(time.Duration(sys.L2Chain.Escape().RollupConfig().BlockTime) * time.Second)
+	for range ticker.C {
+		sequencerUnsafeHead := sys.L2Chain.L2ELNodes()[indexOfSequencer].ChainSyncStatus(sys.L2Chain.Escape().ChainID(), types.LocalUnsafe)
+		t.Logf("Sequencer unsafe head (number %d) is %d/%d blocks ahead of initial unsafe head (number %d)",
+			sequencerUnsafeHead.Number, sequencerUnsafeHead.Number-initialUnsafeHeadNumber, NUM_UNSAFE_BLOCKS, initialUnsafeHeadNumber)
+		if sequencerUnsafeHead.Number >= finalUnsafeHeadNumber {
+			break
+		}
+	}
+
+	// Check verifier nodes progess to the final unsafe head number.
+	allVerifierNodesInSync := func() bool {
+		for i, el := range sys.L2Chain.L2ELNodes() {
+			if i == indexOfSequencer {
+				continue
+			}
+			verifierUnsafeHead := el.ChainSyncStatus(l2chainID, types.LocalUnsafe)
+			if verifierUnsafeHead.Number < initialUnsafeHeadNumber+NUM_UNSAFE_BLOCKS {
+				return false
+			}
+		}
+		return true
+	}
+
+	require.Eventually(t, allVerifierNodesInSync, TIMEOUT, time.Duration(sys.L2Chain.Escape().RollupConfig().BlockTime)*time.Second,
+		"all verifier nodes did not progress to the final unsafe head number within %s",
+		TIMEOUT,
+	)
+	t.Log("All verifier nodes progressed to the final unsafe head number", finalUnsafeHeadNumber)
+
+	// Final sanity check on the block hashes being consistent.
```

Comment:
Is there a possible race condition here? That is, while iterating through the nodes, could the unsafe block hash move beyond that of the sequencer's before we start the loop?